### ISSUE TITLE
Performance drops after PR #1634

### ISSUE THREAD

**Issue by atamazov** [ROCmSoftwarePlatform/MIOpen#1681]
> @junliume This PR fixes Winograd applicability for grouped convolutions; tested on gfx906 using `MIOPEN_DEBUG_FIND_ONLY_SOLVER` environment for the following Winograd solvers: `ConvBinWinogradRxSf2x3` and `ConvBinWinogradRxSf3x2`.
> 
> Was the FindDB that has been used during perf tests generated after this PR? If it was generated before this PR, but after #1533, where bug was introduced, then Winograd solvers may not be in FindDB for grouped convolutions.

Hi @Slimakanzer thank you for getting back to me during your vacation. The following cases have confirmed performance regressions right after this PR (ranked by the impacted lvel):
```
python3 micro_benchmarking_pytorch.py --network inception_v3  --batch-size 128  --iterations 10 --dataparallel --device_ids 0  --fp16 1 && \
python3 micro_benchmarking_pytorch.py --network alexnet  --batch-size 1024  --iterations 10 --dataparallel --device_ids 0  --fp16 1 && \
python3 micro_benchmarking_pytorch.py --network SqueezeNet  --batch-size 128  --iterations 10 --dataparallel --device_ids 0  --fp16 1 && \
python3 micro_benchmarking_pytorch.py --network densenet121  --batch-size 64  --iterations 10 --dataparallel --device_ids 0  --fp16 1 && \
python3 micro_benchmarking_pytorch.py --network alexnet  --batch-size 64  --iterations 10 --dataparallel --device_ids 0  --fp16 0 && \
python3 micro_benchmarking_pytorch.py --network resnet50  --batch-size 128  --iterations 10 --dataparallel --device_ids 0  --fp16 1 && \
python3 micro_benchmarking_pytorch.py --network resnet152  --batch-size 64  --iterations 10 --dataparallel --device_ids 0  --fp16 1 && \
python3 micro_benchmarking_pytorch.py --network vgg16  --batch-size 128  --iterations 10 --dataparallel --device_ids 0  --fp16 1 && \
python3 micro_benchmarking_pytorch.py --network vgg19  --batch-size 128  --iterations 10 --dataparallel --device_ids 0  --fp16 1 && \
python3 micro_benchmarking_pytorch.py --network densenet121  --batch-size 64  --iterations 10 --dataparallel --device_ids 0  --fp16 0
```
I am also quite interested in why they would be impacted by grouped convolutions: did one of the two pre-merge solvers implicitly support it?

CC: @Kirpich30000 @atamazov @JehandadKhan

_Originally posted by @junliume in https://github.com/ROCmSoftwarePlatform/MIOpen/issues/1634#issuecomment-1207665510_

**Comment by atamazov** [comment]
@junliume `urgency_blocker`, `performance`, `bug`.

According to @Slimakanzer's triaging results, the most likely root reason is an issue in #1533 that led to incorrect (unexpected) records in find-db after tuning. W/A to be provided today. Full-blown fix (that should include find-db fixes) a bit later.

**Comment by atamazov** [comment]
@jun Have you seen performance regression on both MI50 (60 CUs) and MI60 (64 CUs)? I am asking because I do not see "unexpected" records in gfx906_60.HIP.fdb.txt, while gfx906_64.HIP.fdb.txt has lots of them. I guess that MI50 is unaffected by the performance regression.

Some stats:
![image](https://user-images.githubusercontent.com/12697412/183886412-73d73f22-2fdb-4edd-ba6c-89da0045f9c4.png)
- "Number of bad" is output of
  - `grep -E ".*-[FBW]=.*ConvBinWinogradRxSf2x3," src/kernels/gfx90878.HIP.fdb.txt | wc -l`.
- "Number of good" is similar but with `ConvBinWinogradRxSf2x3g1`

**Comment by atamazov** [comment]
@JehandadKhan Note that find-db for gfx906/60 and /64 remain unchanged for a long time_** (10 and 14 months respectively). However IIRC we update databases before almost each release. Maybe we update databases only in release branches; if this is the case, then the missing step in the process is merging updated databases from the release branches back to develop (I think the best time for this is immediately after the release). @JehandadKhan What is your opinion?

/cc @junliume 

_moved from https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1682#issuecomment-1210005719_

**Comment by atamazov** [comment]
I did some experiments with Radeon VII (as I do not have MI60 available). PR #1634 does not affect performance on Radeon VII, as expected (because gfx906_60 fdb is good), and, consequently, W/A #1681 has no effect.

In order to simulate MI60-specific regression, I replaced the system find-db with gfx906_64.HIP.fdb.txt. Results of runs with replaced find-db:

  | develop 492eea593 (baseline of W/A) | wa-issue-1681, b868a0eda | W/A adds to performance
-- | -- | -- | --
inception_v3_128_fp16 | 119 | 231 | 48.5%
alexnet_1024_fp16 | 1877 | 2828 | 33.6%
SqueezeNet_128_fp16 | 1073 | 1214 | 11.6%
densenet121_64_fp16 | 345 | 396 | 12.9%
alexnet_64 | 1317 | 1318 | 0.1%
resnet50_128_fp16 | 335 | 350 | 4.3%
resnet152_64_fp16 | 121 | 124 | 2.4%
vgg16_128_fp16 | 146 | 157 | 7.0%
vgg19_128_fp16 | 118 | 125 | 5.6%
densenet121_64 | 267 | 267 | 0.0%

**_The above results match the expectations and confirm the root reason of the issue. Now I can begin implementation of the full-blown fix._**

**Comment by atamazov** [comment]
@junliume This should not be closed because we do not have a full blown fix in develop.

We need #1735 + updated databases (with fixed Winograd 2x3g1/2x3 records) first.

**Comment by atamazov** [comment]
Another issue in the find-db files. According to #1735, ConvBinWinogradRxSf2x3g1 should not be used in WrW (even if n_groups=1). There are some "bad" records:

```
$ for f in src/kernels/*.fdb.txt; do echo ${f} ; grep -E ".*-W=.*ConvBinWinogradRxSf2x3g1," ${f} | wc -l ; done
src/kernels/gfx1030_36.HIP.fdb.txt
11621
src/kernels/gfx803_36.HIP.fdb.txt
0
src/kernels/gfx803_36.OpenCL.fdb.txt
0
src/kernels/gfx803_64.HIP.fdb.txt
0
src/kernels/gfx803_64.OpenCL.fdb.txt
0
src/kernels/gfx900_56.HIP.fdb.txt
5287
src/kernels/gfx900_56.OpenCL.fdb.txt
0
src/kernels/gfx900_64.HIP.fdb.txt
0
src/kernels/gfx900_64.OpenCL.fdb.txt
0
src/kernels/gfx906_60.HIP.fdb.txt
11040
src/kernels/gfx906_60.OpenCL.fdb.txt
0
src/kernels/gfx906_64.HIP.fdb.txt
0
src/kernels/gfx906_64.OpenCL.fdb.txt
0
src/kernels/gfx90878.HIP.fdb.txt
7536
src/kernels/gfx90878.OpenCL.fdb.txt
9956
src/kernels/gfx90a68.HIP.fdb.txt
12079
src/kernels/gfx90a6e.HIP.fdb.txt
11957
```
Current impact: we do not tune Winograd kernels for WrW, so the WrW performance is likely somewhat worse than it can be.

The fix will be included into "full-blown fix" I am developing currently.

/cc @Slimakanzer

**Comment by atamazov** [comment]
Via #1747

**Comment by atamazov** [comment]
Reopened due to:
- https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1841#issuecomment-1307373707
- https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1862#issuecomment-1307394524
- https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1860#issuecomment-1307402103

Some details/explanations: https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1860#issuecomment-1307990033

I am working on the fix.

**Comment by junliume** [comment]
@atamazov  to provide a data point:
Recently we have had some issues with MLIR kernels with large perf discrepancies between tuning and field runs:

*Driver Command*:
```
./bin/MIOpenDriver convfp16 -n 256 -c 64 -H 56 -W 56 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
```
Then we got very interesting observations:
*After #1818*:
```
picked Solution: 103/ConvMlirIgemmFwdXdlops
> Contents found: miopenConvolutionFwdAlgoImplicitGEMM:ConvMlirIgemmFwdXdlops,0.795358,0,miopenConvolutionFwdAlgoImplicitGEMM,not used;miopenConvolutionFwdAlgoWinograd:ConvBinWinogradRxSf2x3g1,1.27296,0,miopenConvolutionFwdAlgoWinograd,not used;miopenConvolutionFwdAlgoDirect:ConvOclDirectFwd,10.4627,0,miopenConvolutionFwdAlgoDirect,not used;miopenConvolutionFwdAlgoGEMM:GemmFwdRest,11.5373,3612672,miopenConvolutionFwdAlgoGEMM,not used
```
Before #1818 
```
picked Solution: 107/ConvAsmImplicitGemmGTCDynamicFwdXdlopsNHWC
> Contents found: miopenConvolutionFwdAlgoImplicitGEMM:ConvAsmImplicitGemmGTCDynamicFwdXdlopsNHWC,0.969281,411115520,miopenConvolutionFwdAlgoImplicitGEMM,not used;miopenConvolutionFwdAlgoWinograd:ConvBinWinogradRxSf2x3g1,1.27296,0,miopenConvolutionFwdAlgoWinograd,not used;miopenConvolutionFwdAlgoDirect:ConvOclDirectFwd,10.4627,0,miopenConvolutionFwdAlgoDirect,not used;miopenConvolutionFwdAlgoGEMM:GemmFwdRest,11.5373,3612672,miopenConvolutionFwdAlgoGEMM,not used
```

However, when I run the actual config, instead of getting `0.795358 ms` as shown above, I am getting `2.076196 ms`
```
MIOpen Forward Conv. Algorithm: 5, Solution: 103/ConvMlirIgemmFwdXdlops
GPU Kernel Time Forward Conv. Elapsed: 2.076196 ms (average)
stats: name, n, c, ho, wo, x, y, k, flopCnt, bytesRead, bytesWritten, GFLOPs, GB/s, timeMs
stats: fwd-conv3x3u1, 256, 64, 56, 56, 3, 3, 64,  59190018048, 102834176, 102760448, 28509, 99, 2.076196
Forward Convolution Verifies OK on CPU reference (0.000196078)
```

**MLIR kernel is taking 0.795 msec in tuning for this config, but actually took 2.076 msec in field run**

**Comment by atamazov** [comment]
@junliume Thanks. The MLIR related issue is not related to this ticket.

**Comment by atamazov** [comment]
@junliume I recommend opening separate ticket for MLIR issue.

**Comment by junliume** [comment]
> @junliume I recommend opening separate ticket for MLIR issue.

Thanks, yes I did in a private repo to MLIR

**Comment by atamazov** [comment]
https://github.com/ROCmSoftwarePlatform/MIOpen/issues/1681#issuecomment-1307404864 resolved by #1870
